File: vbm4dtest.py
import cv2
import numpy as np
import time
import os
import matlab.engine
from PIL import Image

eng = matlab.engine.start_matlab()
start_time = time.time()

# 文件夹路径
noised_folder = 'noised000var100'
output_folder = '000_vbm4d_var100'
os.makedirs(output_folder, exist_ok=True)

# 参数
num_images = 5
varn = 100
image_arrays = []
# 遍历图片文件
for i in range(0, num_images):
    # 构造文件名
    filename = f'{i:08d}.png'

    noised_path = os.path.join(noised_folder, filename)
    current_frame = np.array(cv2.imread(noised_path), dtype=np.float32)

    # 检查图片是否被成功加载
    if current_frame is None:
        print(f"无法读取图像文件 {filename}")
        continue

    image_arrays.append(current_frame)

sigma = varn**0.5  # Noise standard deviation. it should be in the same
# intensity range of the video
profile = 'lc'  # V-BM4D parameter profile
#  'lc' --> low complexity
#  'np' --> normal profile
do_wiener = 1  # Wiener filtering
#   1 --> enable Wiener filtering
#   0 --> disable Wiener filtering
sharpen = 1  # Sharpening
#   1 --> disable sharpening
#  >1 --> enable sharpening
deflicker = 1  # Deflickering
#   1 --> disable deflickering
#  <1 --> enable deflickering
verbose = 1  # 是否输出

images = np.stack(image_arrays, axis=-1)
denoised_images = eng.vbm4d(images, sigma, profile, do_wiener, sharpen, deflicker, verbose)

for i in range(denoised_images.size[3]):
    # 提取第i张图像
    image_array = denoised_images[:][:][:][ i]

    # 将numpy数组转换回PIL图像
    image = Image.fromarray(image_array.astype('uint8'))

    # 保存图像
    save_file_path = os.path.join(output_folder, f'{i:08d}.png')
    image.save(save_file_path)

File: vbm4dtest.m
clear all;close all;

varn = 100;
clean_folder = '000';
noised_folder = 'noised000var100'; % 图片所在文件夹
output_folder = '000_vbm4d_var100_est';
num_images = 100; % 图片数量
% 假设图片大小为 MxN，这里需要根据实际情况调整
M = 720; % 图片的高度
N = 1280; % 图片的宽度
% 初始化数组来存储图片，第四维是图片的数量
images = zeros(M, N, 3, num_images, 'single');

filename = fullfile(clean_folder, sprintf('%08d.png', 0));
% 读取图片
img = imread(filename);
% 存储图片
images(:, :, :, 1) = img;

% 循环读取每张图片
for i = 1:(num_images-1)
    % 构造文件名
    filename = fullfile(noised_folder, sprintf('%08d.png', i));
    % 读取图片
    img = imread(filename);
    % 存储图片
    images(:, :, :, i+1) = img;
end


% Modifiable parameters
sigma = -1;      % Noise standard deviation. it should be in the same
% intensity range of the video
profile = 'np';      % V-BM4D parameter profile
%  'lc' --> low complexity
%  'np' --> normal profile
do_wiener = 1;       % Wiener filtering
%   1 --> enable Wiener filtering
%   0 --> disable Wiener filtering
sharpen = 1;         % Sharpening
%   1 --> disable sharpening
%  >1 --> enable sharpening
deflicker = 1;       % Deflickering
%   1 --> disable deflickering
%  <1 --> enable deflickering
verbose = 1;         % Verbose mode


% V-BM4D filtering
tic;
S = 255;
images = cast(images, 'single')/S;
sigma = sigma/S;
denoised_images = vbm4d(images, sigma, profile, do_wiener, sharpen, deflicker, verbose );
time = toc;

% 循环遍历并保存每张图片
for i = 1:num_images
    % 提取单张图片
    img = denoised_images(:, :, :, i);

    % 构建文件名
    filename = fullfile(output_folder, sprintf('%08d.png', i-1));

    % 保存图片
    imwrite(img, filename);
end







